torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -460,10 +465,22 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
 }
 
 void ProcessGroupAgent::listenLoop() {
-  while (true) {
+  while (rpcRunning_.load()) {
     // rank, tensor size, message type
     std::vector<torch::Tensor> preamble = {torch::empty({3}, {torch::kInt64})};
-    pg_->recvAnysource(preamble, pg_->getRank())->wait();
+    {
+      std::lock_guard<std::mutex> guard(recvWorkMutex_);
+      recvWork_ = pg_->recvAnysource(preamble, pg_->getRank());
```

Comment:
why do you need to hold the mutex across pg_->recvAnySource()?

Reply:
We want to make sure that `recvWork` is in a valid initialized state when we call `abort` on it from the shutdown thread, so we'd want the lock to ensure there is no race when it is initialized here and when it is checked in the `shutdown` function.

Reply:
Ah, I see - I initially didn't see that this was a separate mutex. Thanks.

fwiw, it's a little sketchy to access recvWork_ on 479 without holding the mutex, you should ideally keep a local ref to the shared_ptr, e.g.

auto work = pg_->recvAnysource(preamble, pg_->getRank());
setPendingRecvWork(work);  // or put inline, just grabs lock and sets recvWork_
if (!rpcRunning_.load() ||
   !work->wait() /* aborted */) {
   return;
}

If nothing else, it's easier to reason about shutdown safety, when we guarantee that recvWork_ is only accessed under the lock.

Reply:
Thanks for calling this out, this is implemented in the latest version.

Reply:
Great!
fwiw, LGTM